feat(RELEASE-1387): use trusted artifacts

scoheb opened the pull request:
## Describe your changes
- in an effort to move away from PVCs, we are trying
  trusted artifacts based on https://github.com/konflux-ci/build-trusted-artifacts/tree/main
- tasks that have been updated now support 2 modes:
  - PVC based workspaces
  - Trusted artifacts with emptyDir workspaces
    
Summary:
- collect-data, reduce-snapshot, apply-mapping now support trusted
  artifacts mode
- pipelines that use apply-mapping have had the workspace used for
  the apply-mapping task renamed to be in line with other tasks.
- CI changes:
   - it now attempts to detect if a Trusted Artifacts based task
     has been changes and runs the tests with trusted artifacts enabled.
  - tests now create an on-local-cluster registry.

requires https://github.com/konflux-ci/release-service/pull/666 for use on real clusters

## Relevant Jira
- [RELEASE-1387](https://issues.redhat.com//browse/RELEASE-1387)

## Checklist before requesting a review
- [X] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [X] My commit message includes `Signed-off-by: My name <email>`
- [X] I have bumped the task/pipeline version string and updated changelog in the relevant README
- [X] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)

scoheb committed:
feat(RELEASE-1387): use trusted artifacts

- in an effort to move away from PVCs, we are trying
  trusted artifacts based on:
  https://github.com/konflux-ci/build-trusted-artifacts/tree/main
- tasks that have been updated now support 2 modes:
  - PVC based workspaces
  - Trusted artifacts with emptyDir workspaces

Summary:
- collect-data, reduce-snapshot, apply-mapping now support trusted
  artifacts mode
- pipelines that use apply-mapping have had the workspace used for
  the apply-mapping task renamed to be in line with other tasks.
- CI changes:
  - it now attempts to detect if a Trusted Artifacts based task
    has been changes and runs the tests with trusted artifacts enabled.
  - tests now create an on-local-cluster registry.

Signed-off-by: Scott Hebert <scoheb@gmail.com>

mmalina commented on .github/scripts/deploy_registry.sh:
So all of the kustomization.yml files and the resources are just for the sake of being able to deploy a local registry in kind for local testing, right? (Either as part of Github Workflow, or you could do it locally in your Kind cluster as well.)

What would be used in production? Ideally some local registry in the cluster, right? Otherwise using some custom quay repository would be much slower.

scoheb replied:
> What would be used in production?

quay.io is currently used by build-definitions. We could ask infra to install a local registry much like we do in our tests.

mmalina replied:
ok

mmalina commented on .github/scripts/test_tekton_tasks.sh:
```diff
@@ -22,6 +22,59 @@ shopt -s nullglob
+then
+  echo "Error: env variable TRUSTED_ARTIFACT_OCI_STORAGE not defined"
+  echo ""
+  echo "It must point to a location in quay where trusted artifacts will be placed."
```
This doesn't have to be in quay, right? In fact, the update that you did to the github workflow uses something else (a locally deployed registry).

scoheb replied:
comment was based on outdated code

mmalina replied:
ok

mmalina commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -22,22 +22,48 @@ spec:
+    - name: use-trusted-artifact
+      args:
+        - use
+        - $(params.SOURCE_DATA_ARTIFACT)=$(workspaces.data.path)/work
```
The main point is to get rid of the shared PV, no? So the task should instead mount an emptyDir somewhere and use that, no?

scoheb replied:
when using trusted artifacts, the workspace used is backed by an empty dir

mmalina replied:
I guess it's related to this, right? https://github.com/konflux-ci/release-service/commit/c92d422f2e17e279ece0c6568af0d358964d4e3d#diff-0aaf102c4b7b7b414d8c88cfbb8f460090202777b7ef0eb28287a25cac584e20R242

Is there some documentation for that? I wonder how it works. Because as far as I know, the scope of an emptyDir is only the given pod. So I wonder how a workspace provided in a PipelineRun can be backed up an emptyDir. Even if it's possible, it would then create a new emptyDir for each task, right? Is that how it works? And so you can just decide inside the task if it's a shared workspace like before or if it's an emptyDir workspace and in that case you will load the data via trusted-artifacts first? Looking at the apply-mapping changes, that seems to be the case.

scoheb replied:
> it would then create a new emptyDir for each task, right? 

correct. when using an emptyDir workspace...no data is shared between tasks.

> And so you can just decide inside the task if it's a shared workspace like before or if it's an emptyDir workspace and in that case you will load the data via trusted-artifacts first? 

exactly. I am simply using the `ociStorage` parameter to help the task decide. See:

- https://github.com/konflux-ci/release-service-catalog/pull/784/files#diff-d1a1a5a625bea28920bfde6737b1dbba8fd64f1891027c698c96d0cd6eb0b8f7R64
- https://github.com/konflux-ci/release-service-catalog/pull/784/files#diff-c3855241855369ce8e153dc0b22a899bd43865d84cf135801f92330df1a45007R56

By using that, we can use a feature in Trusted Artifacts that allows you to **skip** any trusted artifact operations when we are in the PVC mode.

Small caveat...we cannot pass an empty `ociStorage` arg to Trusted Artifacts currently. Therefore, we simply use `empty` as our `deactivated` flag.

See examples from both:
- https://github.com/konflux-ci/release-service-catalog/pull/784/files#diff-e4ed74e9c251241373dfd369568f61ab007d3d7cfacef28eb5ab4b481729b7ffR62
- https://github.com/konflux-ci/release-service-catalog/pull/784/files#diff-d1a1a5a625bea28920bfde6737b1dbba8fd64f1891027c698c96d0cd6eb0b8f7R62

mmalina replied:
I see, thanks. I see you have a bunch of checks for "empty", but the `build-trusted-artifacts` image itself has handling for this special string as well, right?

scoheb replied:
Actually Trusted Artifacts does not handle an empty string "" well as-is...hence the string "empty"

johnbieren commented on tasks/managed/apply-mapping/README.md:
```diff
@@ -25,11 +25,18 @@ You can also expand image labels, e.g. "{{ labels.mylabel }}" -> The value of im
+| sourceDataArtifact      | Location of trusted artifacts to be used to populate data directory                                                         | No       | -             |
+| subdirectory            | Subdirectory inside the workspace to be used                                                                                | No       | -             |
+
+## Changes in 1.10.0
```
I think this definitely warrants going to 2.0.0

scoheb replied:
updated

johnbieren commented on tasks/managed/apply-mapping/README.md:
```diff
@@ -25,11 +25,18 @@ You can also expand image labels, e.g. "{{ labels.mylabel }}" -> The value of im
+| snapshotPath            | Path to the JSON string of the Snapshot spec in the config workspace to apply the mapping to                                | No       | -             |
+| dataPath                | Path to the JSON string of the merged data to use in the data workspace                                                     | No       | -             |
+| failOnEmptyResult       | Fail the task if the resulting snapshot contains zero components                                                            | Yes      | false         |
+| ociStorage              | The OCI repository where the Trusted Artifacts are stored.                                                                  | No       | -             |
```
This and line 34 are the only ones with `.` at the end. Probably don't need it

scoheb replied:
fixed

johnbieren commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -22,22 +22,60 @@ spec:
+    - name: subdirectory
+      description: Subdirectory inside the workspace to be used
+      type: string
+      default: ""
```
The README says this isn't optional and has no default

scoheb replied:
fixed

johnbieren commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -22,22 +22,60 @@ spec:
+        #!/usr/bin/env bash
+        set -eo pipefail
+
+        if [ "$(params.ociStorage)" == "empty" ]; then
```
Should `empty` be the default for this parameter then?

scoheb replied:
fixed

johnbieren commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -353,3 +391,29 @@ spec:
               "${SNAPSHOT_SPEC_FILE}" > /tmp/temp && mv /tmp/temp "${SNAPSHOT_SPEC_FILE}"
             fi
         done
+    - name: create-trusted-artifact
```
The first step says we will skip trusted-artifacts steps if the param is empty, but I don't see any when condition on this step. Is that intended?

scoheb replied:
seems you cannot guard Steps from execution ... only Tasks in a Pipeline

the mechanism for skipping involves creating a `skip file` in the extract or create directory.

johnbieren replied:
Gotcha. I think it may be helpful to add like a `# this step does nothing is a .skip-trusted-artifacts file exists` or something (assuming that is accurate) as that is not at all obvious to me

scoheb replied:
I will add appropriate comments to the StepActions PR #844

johnbieren commented on tasks/managed/apply-mapping/tests/test-apply-mapping-content-gateway-disabled.yaml:
```diff
@@ -54,40 +68,98 @@ spec:
                 ]
               }
               EOF
+          - name: skip-trusted-artifact-operations
```
Why are these steps here? They look the same as the one added in the task

scoheb replied:
seeing as there is no workspace to share data between the `task under test` and the `check-results` task...we need to share via Trusted Artifacts.

johnbieren commented on .github/scripts/deploy_registry.sh:
```diff
@@ -0,0 +1,75 @@
+prepare_docker_config
+
+echo ""
+echo "make sure you set these environment variables"
```
Why not have the script run the exports for you?

scoheb replied:
removed the echo statements

johnbieren commented on .github/scripts/deploy_registry.sh:
```diff
@@ -0,0 +1,75 @@
+
+prepare_docker_config() {
+  authString=$(echo -n "root:root" | base64 -w0)
+  cat > "/tmp/.dockerconfig.json" <<EOF
```
Any consideration of using a `mktemp` here instead of a static file?

scoheb replied:
well, I need to know where the file is for this operation: https://github.com/konflux-ci/release-service-catalog/pull/784/files#diff-d7a3587abb00052131d9b5ae6602c6786f2189d7daa363e550859f89ab620bf3R35

johnbieren replied:
Yeah, but (and this relates to my now resolved comment) couldn't you just call the export in the deploy_registry.sh script instead of test_tekton_tasks.sh? Then you wouldn't need to know where the file is in test_tekton_tasks.sh. It seems like it would be a bit more self contained that way, but it is mostly a nit pick anyways

scoheb replied:
I don't follow this:

> couldn't you just call the export in the deploy_registry.sh script instead of test_tekton_tasks.sh

mmalina replied:
I think the idea would be that `deploy_registry.sh` would create a temp file and export the value as DOCKER_CONFIG. And then `test_tekton_tasks.sh` would consume it. I'm fine with the current approach though.

scoheb replied:
ok thanks. can we close this thread?

mmalina replied:
Fine with me. But Johnny's the one who opened it :)

johnbieren commented on pipelines/managed/rh-advisories/rh-advisories-oci-ta.yaml:
Do we still need this? Since the tasks can run in either mode, I thought we wouldn't need this and would set the pipeline to run in ta mode once all tasks are migrated

scoheb replied:
This is used for integration testing. I removed it.

johnbieren commented on pipelines/managed/rh-advisories/rh-advisories-pvc.yaml:
Pretty much same question as rh-advisories-oci-ta

scoheb replied:
This is used for integration testing. I removed it.

johnbieren commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -22,22 +22,62 @@ spec:
       description: Fail the task if the resulting snapshot contains 0 components
       default: "false"
+    - name: ociStorage
+      description: The OCI repository where the Trusted Artifacts are stored.
```
nit - period at end (already removed in README after other comment)

scoheb replied:
fixed

johnbieren commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -22,22 +22,62 @@ spec:
+      default: "empty"
+    - name: ociArtifactExpiresAfter
+      description: Expiration date for the trusted artifacts created in the
+        OCI repository. An empty string means the artifacts do not expire.
```
nit - period at end (already removed in README after other comment)

scoheb replied:
fixed

johnbieren commented on tasks/managed/apply-mapping/tests/test-apply-mapping-content-gateway-disabled.yaml:
```diff
@@ -10,22 +10,36 @@ spec:
   workspaces:
     - name: tests-workspace
+  params:
+    - name: ociStorage
```
So the way this works is the test script file injects this parameter based on what mode it is being called in, right? Pretty clever

scoheb replied:
> Pretty clever

thanks!

johnbieren commented on tasks/managed/collect-data/collect-data.yaml:
Why doesn't this task need
```
    - name: patch-source-data-artifact-result
      image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
      script: |
        #!/usr/bin/env bash
        set -eu
        # this is needed to skip trusted-artifacts tasks
        # when using PVC based workspaces.
        if [ "$(params.ociStorage)" == "empty" ]; then
          echo -n "$(params.ociStorage)" > "$(results.sourceDataArtifact.path)"
        fi
```
?

scoheb replied:
This made me re-look at why we needed this step in the first place. It turns out after iterating in this PR and altering default values, it turns out we no longer need this step in tasks. They are only needed in tests.

I have refactored the use of this step in the StepActions PR #844

johnbieren commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -22,22 +22,62 @@ spec:
+    - name: sourceDataArtifact
+      type: string
+      default: ""
+    - name: subdirectory
```
Can you remind me why this task needs this param when collect-data already outputs each file in a subdir?

scoheb replied:
it has to do with how trusted artifacts are created and then used.

- TA uses the contents of `$(params.dataDir)/$(params.subdirectory)` to create the artifacts
  - let's say `/var/workdir/46ca7138-0b4d-11f0-bc7f-047bcb5d0233/` and archive `*` from that folder.
- So the contents of the artifacts will show files such as `/data.json` without any parent folder
- When the next task wants to use that artifacts, we need to tell it where to extract it to.
  -  So we tell it `$(params.dataDir)/$(params.subdirectory)`
- and then the files will be extracted into `$(params.dataDir)/$(params.subdirectory)` and the task will find `/var/workdir/46ca7138-0b4d-11f0-bc7f-047bcb5d0233/data.json`

Note also that we had some problems trying to extract DIRECTLY into a volume ... i.e. `/var/workdir` ... hence we had to add another subfolder.

jinqi7 commented on tasks/managed/collect-data/collect-data.yaml:
```diff
@@ -72,6 +81,12 @@ spec:
     - name: releasePipelineMetadata
       type: string
       description: json object containing git resolver metadata about the running release pipeline
+    - description: Produced trusted data artifact
```
It's better to use the same attribute order with the previous params.

scoheb replied:
fixed

happybhati reviewed (commented):
LGTM! Huge thanks to @scoheb for not only implementing the changes so neatly but also for the clear explanation and ensuring backward compatibility. Learned a lot from this!

mmalina commented on .github/scripts/test_tekton_tasks.sh:
```diff
@@ -116,7 +136,13 @@ do
+    if [ -z "${USE_TRUSTED_ARTIFACTS}" ]; then
+      workSpaceParams="volumeClaimTemplateFile=$WORKSPACE_TEMPLATE"
+    else
+      workSpaceParams="emptyDir="""
```
I don't think the `""` at the end make a difference. When bash evaluates this, it will just see one string `emptyDir=` plus an empty string which together is still just `emptyDir=`. So:
```suggestion
      workSpaceParams="emptyDir="
```

scoheb replied:
fixed

mmalina commented on .github/workflows/tekton_task_tests.yaml:
```diff
@@ -22,8 +22,26 @@ jobs:
+            taskFile="${dir}/${baseDir}.yaml"
+            if [ -f "${taskFile}" ]; then
+              ociStageParam=$(yq '.spec.params[] | select(.name == "ociStorage") | .name' "${taskFile}")
+              if [ ! -z "${ociStageParam}" ]; then
```
For the opposite of `-z` there is `-n`:
```suggestion
              if [ -n "${ociStageParam}" ]; then
```

scoheb replied:
fixed

mmalina commented on .github/workflows/tekton_task_tests.yaml:
```diff
@@ -22,8 +22,26 @@ jobs:
+          for dir in ${{ steps.changed-dirs.outputs.all_changed_files }}
+          do
+            echo "$dir"
+            baseDir=$(basename "${dir}")
```
Maybe call this `taskName` instead of `baseDir` - makes it more obvious what it is.

scoheb replied:
fixed

mmalina commented on .github/workflows/tekton_task_tests.yaml:
```diff
@@ -22,8 +22,26 @@ jobs:
+            echo "$dir"
+            baseDir=$(basename "${dir}")
+            taskFile="${dir}/${baseDir}.yaml"
+            if [ -f "${taskFile}" ]; then
+              ociStageParam=$(yq '.spec.params[] | select(.name == "ociStorage") | .name' "${taskFile}")
+              if [ ! -z "${ociStageParam}" ]; then
+                echo "Found a trusted artifacts compatible task: ${taskFile}"
+                trustedArtifactsBasedTasks+=($dir)
+              fi
+            fi
```
Maybe this is better? If you prefer yours that's fine.
```suggestion
            if [ -f "${taskFile}" ] \
                && [[ $(yq '[.spec.params[] | select(.name == "ociStorage")] | length' \
                  tasks/managed/apply-mapping/apply-mapping.yaml) -gt 0 ]]
            then
              echo "Found a trusted artifacts compatible task: ${taskFile}"
              trustedArtifactsBasedTasks+=($dir)
            fi
```

scoheb replied:
used yours. thanks

mmalina commented on .github/workflows/tekton_task_tests.yaml:
```diff
@@ -22,8 +22,26 @@ jobs:
+            fi
+          done
+          echo "trustedArtifactsBasedTasks: ${trustedArtifactsBasedTasks[@]}"
+          echo "trustedArtifactsBasedTasks=${trustedArtifactsBasedTasks[@]}" >> $GITHUB_OUTPUT
```
Just to be clear, we assume that if the task has the `ociStorage` parameter, the tests were also updated to support this, right?

scoheb replied:
correct

mmalina commented on CONTRIBUTING.md:
I think this should elaborate more on the two options - running tests either with PVC, or using oci artifacts. And also that not all tasks have been converted yet, so this is only supported for some tasks. (BTW, what will happen if you try to run tasks tests with the env vars set for a task that hasn't been converted? I think it will fail, right? Because only an emptyDir will be provided to the PLR...)

scoheb replied:
I added an intro section about Trusted Artifacts and updated the `env var` section.

scoheb commented:
/retest

jinqi7 commented on .github/scripts/deploy_registry.sh:
```diff
@@ -0,0 +1,68 @@
+deploy_trust_manager() {
+  kubectl apply -k "${script_path}/../resources/trust-manager"
+  sleep 5
+  retry "kubectl wait --for=condition=Ready --timeout=60s -l app.kubernetes.io/instance=trust-manager -n cert-manager pod" \
```
"-n cert-manager"  should be "-n  trust-manager"?

scoheb replied:
nice catch! fixed

scoheb replied:
actually ... trust manager is running in the cert-manager namespace. I reverted

scoheb commented:
quay 502

/retest

scoheb commented:
I have addressed the comments in the PR.

I have now switched to #844 which uses **StepActions** and **features the entire rh-advisories pipeline using Trusted Artifacts**.

superseded by #844
